src/dftbp/dftbplus/mainapi.F90

```diff
@@ -92,6 +96,184 @@ subroutine setGeometry(env, main, coords, latVecs, coordOrigin)
   end subroutine setGeometry
 
 
+  !> Explicitly set the neighbour list instead of calculating it in DFTB+
+  subroutine setNeighbourList(env, main, nNeighbour, iNeighbour, neighDist, cutOff,&
+      & coord, img2CentCell)
+
+    !> Instance
+    type(TEnvironment), intent(inout) :: env
+
+    !> Instance
+    type(TDftbPlusMain), target, intent(inout) :: main
+
+    !> number of neighbours for each atom
+    integer, intent(in) :: nNeighbour(:)
+
+    !> references to image atoms
+    integer, intent(in) :: iNeighbour(:,:)
+
+    !> distances to image atoms
+    real(dp), intent(in) :: neighDist(:,:)
+
+    !> cutoff used for this neighbour list
+    real(dp), intent(in) :: cutOff
+
+    !> coordinates of all images (= atoms in other cells)
+    real(dp), intent(in) :: coord(:,:)
+
+    !> mapping between image index (other cell) and atom index (central cell)
+    integer, intent(in) :: img2CentCell(:)
+
+    real(dp), allocatable :: dist2(:)
+    real(dp), pointer :: rCellVec(:)
+    integer :: nMaxNeighbours, nCellVec, info
+    integer :: iAtom, iCell, iNeigh, iImage, iCellVec, iAtFirst, iAtLast
+    integer, allocatable :: indx(:)
+    logical :: copyData
+
+    if (allocated(main%electronDynamics)) then
+      call error("Not implemented: Cannot set the neighbour list&
+          & when time propagation is enabled")
+    end if
+    if (main%tLocalCurrents) then
+      call error("Not implemented: Cannot set the neighbour list&
+          & when local bond-currents should be computed")
+    end if
+
+    nMaxNeighbours = maxval(nNeighbour)
+
+    main%neighbourList%setExternally = .true.
+
+    #:if WITH_MPI
+      call allocateNeighbourArrays(main%neighbourList, nMaxNeighbours, main%nAtom, .true., env)
+    #:else
+      call allocateNeighbourArrays(main%neighbourList, nMaxNeighbours, main%nAtom, .false.)
+    #:endif
+
+    @:ASSERT(size(nNeighbour) == main%nAtom)
+    main%neighbourList%nNeighbour(:) = nNeighbour(:)
+    main%neighbourList%cutoff = cutOff
+
+    main%nAllAtom = size(img2CentCell) + main%nAtom
+    @:ASSERT(size(img2CentCell) == size(coord, dim=2))
+
+    if (size(main%img2CentCell) /= main%nAllAtom) then
+      call reallocateArrays1(main%img2CentCell, main%iCellVec, main%coord, main%nAllAtom)
+    end if
+
+    !> Prepend data for the atoms in the central cell
+    do iAtom = 1, main%nAtom
+      main%coord(1:3,iAtom) = main%coord0(1:3,iAtom)
+      main%img2CentCell(iAtom) = iAtom
+      main%iCellVec(iAtom) = 1
+    end do
+
+    if (main%nAtom < main%nAllAtom) then
+      main%coord(1:3,main%nAtom+1:) = coord(1:3,:)
+      main%img2CentCell(main%nAtom+1:) = img2CentCell(:)
```

Comment:
I often add those `(:)`, because you can directly understand that an array is assigned by looking at the source. But I do understand that is is convention to remove them im DFTB+.